src/expression_builder.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    execution_context::{ExecutionContext, HELD_VALUE_LOCATION},
+    instruction::Instruction,
+    operators::Operator,
+    TypeSystem,
+};
+
+#[derive(Clone)]
+pub enum Operand<TS: TypeSystem> {
+    Function {
+        addr: usize,
+        args: Vec<Operand<TS>>,
+        stack_size: usize,
+    },
+    ValueRef(usize),
+    ValueRaw(TS::Value),
+    Expression(Box<ExpressionBuilder<TS>>),
+}
+
+#[derive(Default, Clone)]
+pub struct ExpressionBuilder<TS: TypeSystem> {
+    operator: Option<Operator<TS>>,
+    operands: (Option<Operand<TS>>, Option<Operand<TS>>),
+}
+
+impl<TS: TypeSystem> ExpressionBuilder<TS> {
+    pub fn new() -> ExpressionBuilder<TS> {
+        ExpressionBuilder {
+            operator: None,
+            operands: (None, None),
+        }
+    }
+
+    pub fn set_value(&mut self, value: Operand<TS>) -> &mut ExpressionBuilder<TS> {
+        match &self.operands {
+            (None, _) => self.operands.0 = Some(value),
+            (Some(_), None) => self.operands.1 = Some(value),
+            _ => (),
+        }
+        self
+    }
+
+    pub fn set_left_operand(&mut self, value: Operand<TS>) -> &mut ExpressionBuilder<TS> {
+        self.operands.0 = Some(value);
+        self
+    }
+
+    pub fn set_right_operand(&mut self, value: Operand<TS>) -> &mut ExpressionBuilder<TS> {
+        self.operands.1 = Some(value);
+        self
+    }
+
+    pub fn set_operator(&mut self, operator: Operator<TS>) -> &mut ExpressionBuilder<TS> {
+        self.operator = Some(operator);
+        self
+    }
+
+    fn build_function(
```

Comment:
Better specify that it's building the instructions for a function invocation - maybe rename to `build_function_invoke`?

Reply:
how about `expand_function_instructions`

Reply:
That name is a lot harder to understand for me, I feel like `build_function_invoke` or `build_function_invocation` better describes what the function does

Reply:
invoke makes it sound like it is calling the function there

Reply:
are you saing that because the the thing being passed is a function call and not a pointer to a function?

Reply:
#### NOTE
agreed to use the term `funcionCall` over referring to things as *invoke* or *function invoke*